cpp/src/arrow/array/array_test.cc

```diff
@@ -685,6 +685,19 @@ TEST_F(TestArray, TestMakeArrayFromScalarSliced) {
   }
 }
 
+TEST_F(TestArray, TestMakeArrayFromScalarOverflow) {
+  auto scalar = std::make_shared<StringScalar>("aa");
+
+  // Use a length that will cause an overflow when multiplied by the size of the string
+  int64_t length = static_cast<int64_t>(std::numeric_limits<int64_t>::max()) / 2 + 1;
```

Comment:
We should probably test `StringArray` with `int32_t::max` and `LargeStringArray` with `int64_t::max`